blog/2018/12/12/dancer-and-minion/index.markdown

```diff
@@ -0,0 +1,299 @@
+reap the completed process. 
+
+There have been several problems with this approach:
+- it's fragile
+- it doesn't scale
+- it's too easy to screw something up as a developer
+
+In 2019, we have to ramp up to take on a significantly larger workload. The current solution simply will not handle the amount of work we anticipate needing to handle. Enter Minion.
+
+*Note:* The techniques used in this article work equally well with Dancer or Dancer2.
+
+---
+
+## Why Minion?
+
+We looked at several alternatives to Minion, including [beanstalkd](https://beanstalkd.github.io/) and [celeryd](http://www.celeryproject.org/). Using either one of these meant involving our already over-taxed
+infrastructure team, however; using Minion allowed us to use expertise that my team already has without having to burden someone else with assisting us. From a development standpoint, using a product that
+was developed in Perl gave us the quickest time to implementation. 
+
+Scaling our existing setup was near impossible. It's not only not easy to get a handle on what resources are consumed by processes we've forked, but it was impossible to run the jobs on more than one server. 
+Starting over with Minion also gave us a much needed opportunity to clean up some code in sore need of refactoring. With a minimal amount of work, we were able to clean up our XML rendering code and make it work
+from Minion. This cleanup allowed us to more easily get information as to how much memory and CPU was consumed by an XML rendering job. This information is vital for us in planning future capacity.
+
+## Accessing Minion
+
+Since we are a Dancer shop, and not Mojolicious, a lot of convenience you'd get from Mojolicious for working with Minion isn't as available to us. Given we are also sharing some Minion-based
+code with our business models, we had to build some of our own plumbing around Minion:
+
+```
+package MyJob::JobQueue;
+
+use Moose;
+use Minion;
+
+use MyJob::Models::FooBar;
+with 'MyJob::Roles::ConfigReader';
+
+my @QUEUE_TYPES = qw( default InstantXML PayrollXML ChangeRequest );
+
+has 'runner' => (
+    is      => 'ro',
+    isa     => 'Minion',
+    lazy    => 1,
+    default => sub( $self ) {
+        $ENV{ MOJO_PUBSUB_EXPERIMENTAL } = 1;
+        Minion->new( mysql => MyJob::DBConnectionManager->new->get_connection_uri({ 
+            db_type => 'feeds', 
+            io_type => 'rw',
+        }));
+    },
+);
+
+sub has_invalid_queues( $self, @queues ) {
+    return 1 if $self->get_invalid_queues( @queues );
+    return 0;
+}
+
+sub get_invalid_queues( $self, @queues ) {
+    my %queue_map;
+    @queue_map{ @QUEUE_TYPES } = (); 
+    my @invalid_queues = grep !exists $queue_map{ $_ }, @queues;
+    return @invalid_queues;
+}
+
+sub run_job( $self, $args ) {
```

Comment:
I wouldn't call this `run_job` as it doesn't, in fact, run the job. Whether you call it that at $work is up to you but for newcomers, semantic naming is probably better. `queue_job` or `add_job` or something?

Reply:
Changed.